runtime/interpreter/decode.go

```diff
@@ -1008,20 +1162,32 @@ func (d *DecoderV4) decodePath(v interface{}) (PathValue, error) {
 	}, nil
 }
 
-func (d *DecoderV4) decodeCapability(v interface{}) (CapabilityValue, error) {
-	encoded, ok := v.(cborArray)
+func (d *DecoderV4) decodeCapability() (CapabilityValue, error) {
+
 	const expectedLength = encodedCapabilityValueLength
-	if !ok || len(encoded) != expectedLength {
-		return CapabilityValue{}, fmt.Errorf("invalid path encoding: expected [%d]interface{}, got %T",
+
+	size, err := d.decoder.DecodeArrayHead()
+	if err != nil {
+		if e, ok := err.(*cbor.WrongTypeError); ok {
+			return CapabilityValue{}, fmt.Errorf("invalid capability encoding: expected [%d]interface{}, got %s",
+				expectedLength,
+				e.ActualType.String(),
+			)
+		}
+		return CapabilityValue{}, err
+	}
+
+	if size != expectedLength {
+		return CapabilityValue{}, fmt.Errorf("invalid capability encoding: expected [%d]interface{}, got [%d]interface{}",
 			expectedLength,
-			v,
+			size,
 		)
 	}
 
 	// address
 
-	field1 := encoded[encodedCapabilityValueAddressFieldKey]
-	field1Value, err := d.decodeValue(field1, nil)
+	// Decode address at array index encodedCapabilityValueAddressFieldKey
+	field1Value, err := d.decodeValue(nil)
 	if err != nil {
 		return CapabilityValue{}, fmt.Errorf("invalid capability address: %w", err)
```

Comment:
Just realized: Could we directly call `decodeAddress` here? This would avoid the type dispatch in `decodeValue` and the type check below

Reply:
Great idea!